pkg/reporterreporter/report.go

```diff
@@ -30,49 +38,98 @@ type RequestData struct {
+		http.StatusRequestTimeout:  true,
+		http.StatusTooManyRequests: true,
+	}[statusCode]
+
+	if ok {
+		return true
+	}
+
+	return statusCode >= 500 && statusCode <= 599
+}
+
+func (r *Report) sendData(requestData RequestData) (retriable bool, err error) {
+	req, err := r.mutator.NewRequest(restclient.BodyFromJSON(requestData))
+	if err != nil {
+		return false, errors.Wrap(err, "could not craft a HTTP request")
+	}
+
+	resp, err := r.HTTPClient.Do(req)
+	if err != nil {
+		return true, errors.Wrap(err, "failed to perform HTTP request to slurper")
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		return false, nil
+	}
+
+	if isRetriableError(resp.StatusCode) {
+		return true, errors.Errorf("POSTing data to slurper failed with retriable status code %d", resp.StatusCode)
+	}
+
+	return false, errors.Errorf("POSTing data to slurper failed with status code %d", resp.StatusCode)
+}
+
 func (r *Report) Run(ctx context.Context) error {
 	namespaces, err := r.KubernetesClient.CoreV1().Namespaces().List(meta_v1.ListOptions{})
 	if err != nil {
 		return errors.Wrap(err, "failed to list namespaces")
 	}
+
 	for _, namespace := range namespaces.Items {
-		list, err := r.ReporterClient.ReporterV1().Reports(namespace.Name).List(meta_v1.ListOptions{})
+		reports, err := r.ReporterClient.ReporterV1().Reports(namespace.Name).List(meta_v1.ListOptions{})
 		if err != nil {
-			return errors.Wrap(err, "could not list reports")
+			r.Logger.Error("Could not list reports in namespace", zap.String("namespace", namespace.Name), zap.Error(err))
+			continue
 		}
-		if len(list.Items) == 0 {
-			r.Logger.Sugar().Infof("%q is empty", namespace.Name)
+
+		if len(reports.Items) == 0 {
+			r.Logger.Info("Report for namespace is empty", zap.String("namespace", namespace.Name))
 			continue
 		}
 
-		for _, report := range list.Items {
+		for _, report := range reports.Items {
 			requestData := RequestData{
 				Cluster:   r.Cluster,
 				Service:   namespace.Name,
 				Timestamp: time.Now().Unix(),
 				Data:      report.Report,
 			}
 
-			rm := restclient.NewRequestMutator(
-				restclient.BaseURL(r.RemoteURI),
-			)
-			req, err := rm.NewRequest(
-				restclient.JoinPath("/slurp"),
-				restclient.Method(http.MethodPost),
-				restclient.BodyFromJSON(requestData),
-			)
-			if err != nil {
-				return errors.Wrap(err, "invalid request")
-			}
-			resp, err := r.HTTPClient.Do(req)
-			if err != nil {
-				return errors.Wrap(err, "failed to connect to backend")
-			}
-			if resp.StatusCode != http.StatusOK {
-				return errors.New("failed to POST the data")
+			for i := 1; i <= retryAttemptsPerReport; i++ {
+				if retriable, err := r.sendData(requestData); err != nil {
+					if !retriable {
+						// If the request is not retriable, then we shouldn't try the other requests either
+						return err
+					} else {
+						r.Logger.Warn("Failed to send request data to slurper",
+							zap.Int("attempt", i),
+							zap.Int("max_attempts", retryAttemptsPerReport),
```

Comment:
I'd actually be okay with not having these fields. I think it's nice to restrict these to fields we actually want to filter on because if we use splunk the most common way is to look at the msg field and use the json fields as filtering (or have error/msg as two separate columns).

Plus the more fields we have here, the more clog we get in splunk